packages/legend-query/src/stores/QueryBuilderMilestoningHelper.ts

```diff
@@ -0,0 +1,615 @@
+      ) &&
+      !prevExpression.derivedProperty.name.endsWith(
+        MilestoneVersionPropertySufixes.ALL_VERSIONS_IN_RANGE,
+      ) &&
+      prevExpression.parameterValues.every(
+        (p, index) => p === currentExpression.parameterValues[index],
+      )
+    ) {
+      currentExpression.propertyExpression.parametersValues = [
+        guaranteeNonNullable(
+          currentExpression.propertyExpression.parametersValues[0],
+        ),
+      ];
+    }
+    prevExpression = currentExpression;
+  }
+};
+
+export const fillMilestonedDerivedPropertyArguments = (
+  derivedPropertyExpressionState: QueryBuilderDerivedPropertyExpressionState,
+  temporalTarget: MILESTONING_STEROTYPES,
+  idx: number,
+): ValueSpecification | undefined => {
+  const querySetupState =
+    derivedPropertyExpressionState.queryBuilderState.querySetupState;
+  switch (temporalTarget) {
+    case MILESTONING_STEROTYPES.BUSINESS_TEMPORAL: {
+      if (!querySetupState.businessDate) {
+        querySetupState.setBusinessDate(
+          derivedPropertyExpressionState.queryBuilderState.buildMilestoningParameter(
+            DEFAULT_MILESTONING_PARAMETERS.BUSINESS_DATE,
+          ),
+        );
+      }
+      const parameter = querySetupState.businessDate;
+      derivedPropertyExpressionState.businessDate =
+        querySetupState.BusinessDate;
+      return parameter;
+    }
+    case MILESTONING_STEROTYPES.BITEMPORAL: {
+      if (!querySetupState.processingDate) {
+        querySetupState.setProcessingDate(
+          derivedPropertyExpressionState.queryBuilderState.buildMilestoningParameter(
+            DEFAULT_MILESTONING_PARAMETERS.PROCESSING_DATE,
+          ),
+        );
+      }
+      if (!querySetupState.businessDate) {
+        querySetupState.setBusinessDate(
+          derivedPropertyExpressionState.queryBuilderState.buildMilestoningParameter(
+            DEFAULT_MILESTONING_PARAMETERS.BUSINESS_DATE,
+          ),
+        );
+      }
+      if (idx === 0) {
+        derivedPropertyExpressionState.processingDate =
+          querySetupState.ProcessingDate;
+        return querySetupState.processingDate;
+      } else {
+        derivedPropertyExpressionState.businessDate =
+          querySetupState.BusinessDate;
+        return querySetupState.businessDate;
+      }
+    }
+    case MILESTONING_STEROTYPES.PROCESSING_TEMPORAL: {
+      if (!querySetupState.processingDate) {
+        querySetupState.setProcessingDate(
+          derivedPropertyExpressionState.queryBuilderState.buildMilestoningParameter(
+            DEFAULT_MILESTONING_PARAMETERS.PROCESSING_DATE,
+          ),
+        );
+      }
+      const parameter = querySetupState.processingDate;
+      derivedPropertyExpressionState.processingDate =
+        derivedPropertyExpressionState.queryBuilderState.querySetupState.ProcessingDate;
+      return parameter;
+    }
+    default:
+      return undefined;
+  }
+};
+
+export const isValidMilestoningLambda = (
```

Comment:
Maybe, let's rename this to `processMilestoning...` is `Lambda` the right word here? Also, if you name the parameter `propertyExpression`, could we narrow down the type somehow? Due to its awkwardness to name and define this function, I think maybe instead of calling this on `propertyExpression.parametersValues[0]`, why don't we call it on `propertyExpression` itself, that way, we can name this function something like `processMilestoningPropertyExpression` instead. What do you think?

Reply:
Makes more sense